### ISSUE TITLE
PetaPoco Paging not working correctly

### ISSUE THREAD

**Issue by fabianabela** [CollaboratingPlatypus/PetaPoco#349]
Hi Everyone,

I am using PetaPoco (the one that ships with Umbraco) and have a scenario where I have paging, thus I decided to make use of the Page method. The issue is that it is not working correctly at all. 

I have done some debugging and noticed that with a very simple query, like for example: 

SELECT * 
FROM property 
WHERE City='New York' 

it works ok but then as you feed a little bit of a more complex query like for example: 

SELECT * 
FROM property 
JOIN images ON images.PropertyId = property.ID
WHERE City='New York'
GROUP BY property.ID
ORDER BY property.LastUpdated

it just returns wrong results, like for example it returns 39 items but then the TotalItems value is 33. To be honest I tried but can't really figure out what is the issue. Is this method that delecate? Am I missing something?

Can someone please point me to the right direction? Thanks a lot.

**Comment by pleb** [comment]
`JOIN images ON images.PropertyId = property.ID` I bet this is causes the extra return rows. Do you have any records where there is more than 1 image per property?

**Comment by fabianabela** [comment]
@pleb of course there are records where there is more than 1 image per property, but that is where the Group by comes in, I am basically grouping the results with the property id so that only 1 record per property is returned, thus I will have no duplicate properties. Basically then in the select statement of my actual query I am concatenating all the image file names in one string, but I have left that out here so to keep it simple.

I am not sure what the issue is for instance I ran a test case where I pass to the page method a query that returns 39 rows, current page = 1 and items per page = 100. Thus there should only be one total page and the TotalItems value should be equal to the count of the returned Items list, but still the TotalItems value returned by the page method was invalid.

Is this method not compatible with sql queries that have joins and grouping by any chance?

Thanks a lot for your help, I really appreciate it.

**Comment by pleb** [comment]
Ah sorry. I missed the group by. 😞 

I think I know where it could be going wrong. The query is altered by PP to support paging, with the alternation specific to the underlying DB. Based on the leniency of the GROUP BY I'm going to guess MySQL? 

If you're using the non-compiled version. Add a breakpoint at the end of the `private void BuildPageQueries<T>(long skip, long take, string sql, ref object[] args, out string sqlCount, out string sqlPage)` method and inspect the query. 

Otherwise, you can extend PP and override the `public virtual void OnExecutedCommand(IDbCommand cmd)` and inspect the `cmd.CommandText` property to see what's going on.

Let me know how you go

**Comment by fabianabela** [comment]
@pleb yes it is MySQL, I have done as you instructed and this is the query that was being produced:

SELECT COUNT(*) 
FROM property
INNER JOIN images ON images.PropertyId=property.IntegratorPropertyID
WHERE (City='Zejtun')
GROUP BY property.IntegratorPropertyID

Now the issue with such query is that it returns the count for every group and not the total number of rows returned. That is why the TotalItems was always returning an incorrect value.

In order to fix such issue instead of using the Page method I have made use of the Fetch method in order to handle the limit and offset for me, as follows:

var properties = db.Fetch<PropertyListingItemPartialModel>(currentPageNumber, pageSize, sqlQuery);

Then in order to get the TotalItems I am running another query as follows:

var totalNumberOfProperties = db.SingleOrDefault<int>("SELECT FOUND_ROWS();");

Please note that for the above query to work one needs to add the following: "SQL_CALC_FOUND_ROWS" right after the SELECT statement. Such technique is explained in the accepted answer of this stackoverflow question: http://stackoverflow.com/questions/364825/getting-the-number-of-rows-with-a-group-by-query

So basically I have replaced the Page method with the following:

var properties = db.Fetch<PropertyListingItemPartialModel>(currentPageNumber, pageSize, sqlQuery);   
            var totalNumberOfProperties = db.SingleOrDefault<int>("SELECT FOUND_ROWS();");
                
            var page = new Page<PropertyListingItemPartialModel>()
            {   
                Items = properties,
                TotalItems = totalNumberOfProperties
            };

It would be great if you could fix such issue in a later update so others do not encounter such headache. Besides this issue PetaPoco is great, thank you very much.

**Comment by Ste1io** [comment]
Could @butulia , @asherber , or @pleb confirm if this issue can be closed? It appears to be a duplicate of, or at least related to, #534 (closed by PR #535).
